### DISCUSSION TITLE
iPXE can't download http after reboot from linux, only poweroff

### DISCUSSION THREAD

**Discussion by FallingSnow** [ipxe/ipxe#704] [Q&A] [answered]
So I'm running into a rather strange issue where I can only successfully chainload `default.ipxe` if I completely power off. If I instead reboot from the linux iPXE got me into in the first place, it can no longer connect. I print out the IPv6 address that it gets as well as a gateway and dns so I know that is working. But for some reason it seems like unless the network card is completely powered off then it can't access the http resource. This is the error code it gives me: https://ipxe.org/err/4c0a60

Once I press the reset but on the PC and it boots into iPXE everything is smooth sailing.

Is there anyway to completely reset the network interface?

**Comment by NiKiZe** [comment]
What is default.ipxe? (How was it built?)
Which NIC is it? At least include output of ifstat in iPXE.

Any chance you can tcpdump the traffic?

**Reply by FallingSnow** [reply]
I have a web server serving default.ipxe. default.ipxe is just a script I modified. I'll share it below. It shows the full address it is trying to access when it's trying to connect so I know it gets `${nextserver}` correctly.

I built the ipxe image from the latest git via `make -j11 bin-x86_64-efi/ipxe.efi NO_WERROR=1 EMBED=boot.ipxe TRUST=$TRUST CERT=$CERT`

`boot.ipxe`:
```ipxe
#!ipxe

# see https://ipxe.org/docs
# see https://ipxe.org/settings

# console -x 1920 -y 1080 ||

# Always require signed images
imgtrust --permanent
echo Signed images are required to boot
echo

# show the system information.
# NB you can run "config" to interactively see all the possible variables.
echo Platform..... ${platform}
echo Architecture. ${buildarch}
echo Product...... ${smbios/product}
echo Manufacturer. ${smbios/manufacturer}
echo Asset........ ${smbios/asset}
echo UUID......... ${smbios/uuid}
echo Serial....... ${smbios/serial}

# Prompt to enter shell
prompt --timeout 2000 Press any key for the iPXE command line... && shell ||

# show the available network interfaces.
echo Network interfaces:
ifstat
echo

set idx:int32 0
:loop isset ${net${idx}/mac} || goto loop_done
  echo Configuring net${idx} ${net${idx}/mac} (${net${idx}/chip})...

ifconf net${idx}

isset ${net${idx}/ip} || goto no_ipv4
echo MAC address.. ${net${idx}/mac}
echo IP address... ${net${idx}/ip}
echo Subnet mask.. ${net${idx}/netmask}
echo Gateway...... ${net${idx}/gateway}
echo DNS.......... ${net${idx}/dns}
echo Hostname..... ${hostname}
echo Domain....... ${domain}
echo Next Server.. ${next-server}
echo Filename..... ${filename}
echo
:no_ipv4

isset ${net${idx}/ip6} || goto no_ipv6
echo MAC address.. ${net${idx}/mac}
echo IP address... ${net${idx}/ip6}/${net${idx}/len6}
echo Gateway...... ${net${idx}/gateway6}
echo DNS.......... ${net${idx}/dns6}
echo
:no_ipv6

isset ${net${idx}/ip} || isset ${net${idx}/ip6} || goto next_loop

isset ${next-server} && set nextserver ${next-server} || set example.com
isset ${filename} || set filename /boot/default.ipxe

nslookup syslog_address ${nextserver}
set syslog6 ${syslog_address}
set url http://${nextserver}${filename}?mac=${net${idx}/mac}&uuid=${smbios/uuid}&asset=${smbios/asset}
set verifyurl http://${nextserver}${filename}.sig?mac=${net${idx}/mac}&uuid=${smbios/uuid}&asset=${smbios/asset}

:load_image
echo Loading ${url}
imgload --name next ${url} || goto next_loop
#echo Verifying via ${verifyurl}
imgverify next ${verifyurl} || goto next_loop
imgstat
chain --autofree || goto next_loop

:next_loop
inc idx && goto loop
:loop_done

:fail
prompt --timeout 5000 Press any key for the iPXE command line... && shell ||
menu Failed to load default target. Select action
item --default reboot Reboot
item shell Shell
item recovery Linux Recovery
choose --timeout 10000 --default reboot target
clear menu
goto ${target}

:recovery
echo Not currently supported

:shell
shell

:reboot
reboot

:exit
exit 0
```

NIC: `06:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 15)`
I believe it's using the 6168 driver.

`default.ipxe`:
```ipxe
#!ipxe

:menu
set space:hex 20:20
set space ${space:string}
menu [ Boot Menu ]
item --gap Boot
item alpine ${space} Alpine
item arch ${space} Arch
item --gap Utilities
item shell ${space} iPXE Shell
item reboot ${space} Reboot system
item poweroff ${space} Poweroff system
choose --timeout 4000 --default alpine item
clear menu
goto ${item}

:alpine
set url http://${nextserver}/boot/alpine/alpine.ipxe
goto load_img

:arch
set url http://${nextserver}/boot/arch/arch.ipxe
goto load_img

:load_img
imgselect --name next ${url}?uuid=${smbios/uuid}&mac=${net0/mac}
imgverify next ${url}.sig || goto shell
chain --autofree || goto shell

:shell
echo Type "exit" to return to menu.
shell
goto menu

:reboot
reboot
goto menu_exit

:poweroff
poweroff
goto menu_exit

:menu_exit
exit 0
```

I'll work on getting an `ifstat` output.

A tcpdump might be a bit of a challenge. I don't have any computers that can sit in the middle to dump the traffic. I'm on a laptop with no ethernet ports.

**Reply by NiKiZe** [reply]
Sorry, my mistake on understanding default.ipxe, but you answered my question with ipxe.efi anyway which is awesome.

Yes the rtl81xx is a mess to get corrects resets on. But usually it is totally dead and you can't even get DHCP in those cases. But you at least have RA working?

Could you tell us more on how the NIC is configured in OS before reboot? MTU, vlans, bridged?
When http fails, does it still work for ping or any other smaller packets, or does it go totally dead when that happens? (ifstat should be able to give some details)

**Reply by FallingSnow** [reply]
Ok, I'm all kinds of confused. This is actually kind of complicated but here goes. *Sorry for the wall of text*

I was running into this issue for a while where even if it failed in ipxe for another unrelated reason and rebooted without a complete poweroff it wouldn't be able to connect anymore. So I had just been toggling the power switch. One day that board's NIC just stopped working all together. I updated bios and what not and somehow the board won't even boot anymore, I think if I took out the DP->HDMI->VGA adapter I would have been able to boot. But by the time I realized that the adapter is most likely the issue I already contacted Newegg for an RMA.

So now I know the NICs on this board model is buggy. I had another board (the exact same model) which I was able to boot and NIC worked so I just decided to use that. So everything is fine and dandy until I start seeing this issue I had with the other board popping up (connection timeout) after reboot from OS. So I decided this is happening with another board, time to make a post about this, which brings us to today.

Well, now I can't replicate the issue. So I'm starting to wonder if I'm crazy and was just connecting two issues on two boards that weren't related. But luckily I took a picture of when the issue was happening on this new board (with `DEBUG=tcp`) so I know I'm not.
![Screenshot_20220813-102112_Camera](https://user-images.githubusercontent.com/2512197/184509321-41ffc587-08c4-4cd6-9e87-df725adc700e.png)

Either way I can't seem to replicate it now.

> But you at least have RA working?

Yup. Gets an IP and route and DNS.

> Could you tell us more on how the NIC is configured in OS before reboot? MTU, vlans, bridged?
When http fails, does it still work for ping or any other smaller packets, or does it go totally dead when that happens? (ifstat should be able to give some details)

NIC was just left up to Alpine Linux's default config, I did nothing else to it. I do know that Alpine does do SLAAC again to get IP/Route/DNS.
Pinging it while it tries to connect is a good test, if it happens again I'll have to try it.

I'll let you know if I can replicate the issue.

**Reply by NiKiZe** [reply]
With that info and knowing ping is good then I think tcpdump is needed, you could try this on the webbserver to see if it gets anything from the iPXE client.

**Reply by FallingSnow** [reply]
Ok, it has started happening again. A part of an ipxe script failed and caused it to reboot and now it's stuck in a loop of being unable to connect to webserver where it chainloads the script from.

I've tried pinging from the webserver it chainloads the script from and it can ping it, even while it is trying to connect to the webserver.

I'm not sure I understand what's wrong in the tcp dump. I did it on the webserver side.
[ipxe-connecting.zip](https://github.com/ipxe/ipxe/files/9386052/ipxe-connecting.zip)

**Reply by FallingSnow** [reply]
Okay, I opened a pxe shell after the connection failed and tried pinging the webserver and it worked. Then I tried chain loading the image by writing out the command and it worked...

**Reply by stappersg** [reply]
Regarding the _"and it worked ..."_:
Please check in the access logging of the webserver if it was HTTP over IPv4 or  IPv6.

**Reply by FallingSnow** [reply]
@stappersg The ipxe machine only has IPv6.

**Comment by stappersg** [comment]
> I'm not sure I understand what's wrong in the tcp dump.

As I Understand the `.pcap` file, is iPXE failing to do HTTP over IPv6.

* frame 24:  iPXE  (attemps) starting the TCP stream
* frame 25: webserver replies   (I don't know it is _"all fine for me"_ or _"with some restrictions"_)
* frame 26: iPXE stops the attempt ( sends `RST`)
* frame 27: iPXE retransmits the start attempt as in frame 24
* frame 28: iPXE retransmits the start attempt as in frame 24

I myself do HTTP over IPv4, have no first hand experience with iPXE doing HTTP over IPv6.  Things I would like to see:
* Reports of _"iPXE IPv6  HTTP works for me"_
* Version of iPXE that FallingSnow is using.  ( output of `make version`  ( Yes, I am aware of the _"I built the ipxe image from the latest git"_ ))

**Reply by FallingSnow** [reply]
So it seems like this is an issue for iPXE over HTTP6. Maybe iPXE doesn't like something the server is sending back?
```
$ make version
1.21.1+ (g33cf7)
```

**Reply by stappersg** [reply]
Over here
```text
$ make version
1.21.1+ (gc5af4)
```
Which matches https://github.com/ipxe/ipxe/commit/c5af41a6f5b5f4a420b3e539f9e3a8dc9f8dd03e as latest commit.

And for what it is worth:
```text
$ git log --oneline | grep 33cf7
$ git log --oneline | grep c5af4
c5af41a6 [intelxl] Explicitly request a single queue pair for virtual functions
$ git log --oneline | head -n 3
c5af41a6 [intelxl] Explicitly request a single queue pair for virtual functions
04879352 [intelxl] Allow for admin commands that trigger a VF reset
491c075f [intelxl] Negotiate virtual function API version 1.1
$ 
```
( I'm trying to say I have **no** `commit 33cf7` in my _iPXE git clone_)

> Maybe iPXE doesn't like something the server is sending back?

Yes, that could be.  But the web server isn't doing that every time. Anyway: _Please tell us more about the web server we are dealing with._

**Reply by stappersg** [reply]
Also have something like
```bash
sudo tcpdump  -w attemptNNN.pcap  port http and ether host MA:CA:DD:RA:SU:SR
```
at the webserver where NNN is some counter for next attempts.  The idea is being able to compare _succeses_ and _fails_.

**Reply by stappersg** [reply]
Oops, the
* frame 25: webserver replies   (I don't know it is _"all fine for me"_ or _"with some restrictions"_)

should have been
* frame 25: webserver replies   (I don't know if is _"all fine for me"_ or _"with some restrictions"_)

or
* frame 25: webserver replies   (I don't know it means _"all fine for me"_ or _"with some restrictions"_)

(sorry for the _typo_)

**Reply by FallingSnow** [reply]
> Over here

Oh sorry about that. I forgot I cloned iPXE to save some config changes I made. Looks like my fork is about a month old. It doesn't seem like any of the new commits would fix my issue (except maybe https://github.com/ipxe/ipxe/commit/0965cec53c9e421f900bb050b1c83196a0a5687b). Here is my repo: https://github.com/universalis-llc/ipxe

> The idea is being able to compare succeses and fails.

Would you like a pcap of a successful boot?

**Reply by stappersg** [reply]
> Would you like a pcap of a successful boot?

Yes.

**Reply by FallingSnow** [reply]
Okay, I'll send you one later today.

**Comment by mcb30** [comment, marked as answer]
@FallingSnow From your capture (`ipxe-connecting.zip`), it looks as though the behaviour happens when iPXE reuses a port number (15117) to initiate the HTTP connection on port 80 (packet 24).  This causes the server to respond with an ACK for a sequence number from the previous connection (packet 25).  iPXE then responds with a RST to tear down this old connection (packet 26).  This RST seems to be ignored and the stale connection is never torn down, leaving both iPXE and the server unable to establish a new connection using this pair of port numbers.

From memory (i.e. without rechecking the TCP RFCs), I think iPXE is doing the right thing here.  It is possible that a firewall rule on the server side is causing it to ignore the RST from iPXE.  It would be worth checking if you can reproduce the behaviour with all IPv6 firewall rules cleared on the server side (and input and output policies both set to `ACCEPT`).

**Reply by FallingSnow** [reply]
That packet trace was taken on the web server so a firewall shouldn't be the issue right?

The web server is a [nodejs koa](https://github.com/koajs/koa) server which I assume just uses NodeJS's http implementation under the hood.

This kind of stinks because I think the Koa guys are just gonna tell me it's a node issue and the Node guys are gonna tell me they won't look into it until I test it with just raw NodeJS which is... work -- I guess.

Thanks anyway for looking into it everyone and coming up with the probable cause.

**Reply by stappersg** [reply]
> Thanks anyway for looking into it everyone and coming up with the probable cause.

:blush:  there is now #710  :wink:

**Reply by 9001** [reply]
Just chiming in to say I had the same issue, and that's the story behind pullrequest #626 since it made the problem go away :>

In my case it wasn't related to warmboot/coldboot, instead it would happen on quick successive powercycles. And it didn't matter what webserver I used (tried fastapi+uvicorn and then apache2), which makes sense because *i think* it's the OS/kernel that keeps track of tcp sequencenumbers and ephemeral ports, rather than the httpd?

**Reply by FallingSnow** [reply]
That would make since I've experienced this on 2 different servers too, a NodeJS server and a rust [hyper](https://github.com/hyperium/hyper) based webserver (miniserve).

**Comment by stappersg** [comment]
@FallingSnow which documention are you willing to provide to reproduce your _nodejs koa server setup_?

The idea is having a counter part for #730.

**Reply by stappersg** [reply]
Parking a snippet of `curl (1)` here.
```text
       --local-port <num/range>
              Set  a preferred single number or range (FROM-TO) of lo‐
              cal port numbers to use  for  the  connection(s).   Note
              that  port  numbers by nature are a scarce resource that
              will be busy at times so setting this range to something
              too  narrow  might  cause  unnecessary  connection setup
              failures.

              Example:
               curl --local-port 1000-3000 https://example.com
```

**Comment by FallingSnow** [comment]
Okay. I'm having this very strange situation happen again. This time without a node server. I'm using (miniserve)[https://github.com/svenstaro/miniserve] instead. I ran it in verbose mode and never even saw a request when failing. I've got both a good and bad tcp dump.
[tcpdump.zip](https://github.com/ipxe/ipxe/files/9796072/tcpdump.zip)

 After trying a bunch of things:
- restart switch
- remove haproxy between client and miniserve
- replug server ethernet port and client ethernet port
- reset button on client (multiple times)

I finally powered off the client by pulling the plug, then plugging it back in. This worked right away and it was no longer in a loop trying to connect.

Could this have something to do with the interface not being reset correctly?

**Reply by NiKiZe** [reply]
As mentioned earlier "the rtl81xx is a mess to get corrects resets on." Have you considered running your tests on a virtual machine, such as qemu?

**Reply by FallingSnow** [reply]
That's why I mentioned it. Is there something I can do to determine if the interface is reset correctly? Some debug option that lists the interface states?

> Have you considered running your tests on a virtual machine, such as qemu?

I haven't. I feel like getting this issue to happen in qemu (which I doubt it would) would be hard to do considering it randomly happens on my real hardware.

**Reply by NiKiZe** [reply]
The reason to use qemu is to debug your http server.
You had a server that didn't work with iPXE. You also have a buggy NIC. Trying to deal with both at once creates unreasonable challenge.
